Add add_slice_to_data

wthrowe opened the pull request.

wthrowe committed:
Add add_slice_to_data

osheamonn commented on src/DataStructures/VariablesHelpers.hpp:
```diff
@@ -43,3 +46,32 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
+/*!
+ * \ingroup DataStructures
+ * \brief Adds data on a codimension 1 slice `slice` to a volume
+ * quantity `data`. The slice has a constant grid coordinate in
```
Is it not the case that the slice has constant logical coordinate and not grid coordinate?

kidder commented on src/DataStructures/VariablesHelpers.hpp:
```diff
@@ -11,13 +11,16 @@
 /*!
  * \ingroup DataStructures
  * \brief Slices the data within `vars` to a codimension 1 slice. The
  * slice has a constant grid coordinate in direction `sliced_dim`, and slices
```
this should be logical coordinate as well

kidder commented on src/DataStructures/VariablesHelpers.hpp:
```diff
@@ -43,3 +47,33 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
+ */
+template <std::size_t VolumeDim, typename TagsList>
+void add_slice_to_data(const gsl::not_null<Variables<TagsList>*> data,
+                       const Variables<TagsList>& slice,
```
slice -> vars_on_slice

kidder commented on src/DataStructures/VariablesHelpers.hpp:
```diff
@@ -43,3 +47,33 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
+ * \see data_on_slice
+ */
+template <std::size_t VolumeDim, typename TagsList>
+void add_slice_to_data(const gsl::not_null<Variables<TagsList>*> data,
```
volume_vars

kidder commented on src/DataStructures/VariablesHelpers.hpp:
```diff
@@ -43,3 +47,33 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
+                       const Index<VolumeDim>& extents,
+                       const size_t sliced_dim,
+                       const size_t slice_point) {
+  const size_t number_of_independent_components =
```
get constexpr

kidder commented on src/DataStructures/VariablesHelpers.hpp:
```diff
@@ -11,12 +11,16 @@
- * slice has a constant grid coordinate in direction `sliced_dim`, and slices
- * the volume through the point `slice_point` of this coordinate.
+ * slice has a constant logical coordinate in direction `sliced_dim`,
+ * and slices the volume through the point `slice_point` of this
```
I find "the point `slice_point` of this coordinate" to be confusing as you are not supposed to pass the value of the fixed coordinate, but instead the index that is fixed in `sliced_dim`.  

How about something like:
 "and slices the volume at `fixed_index` in `sliced_dim`.  For example, to get the lower boundary of `sliced_dim`,  pass `0` for `fixed_index`; to get the upper boundary, pass `extents[sliced_dim] - 1`"

Thoughts?

kidder commented on src/DataStructures/VariablesHelpers.hpp:
```diff
@@ -43,3 +47,33 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
+ * \ingroup DataStructures
+ * \brief Adds data on a codimension 1 slice to a volume quantity. The
+ * slice has a constant logical coordinate in direction `sliced_dim`,
+ * and slices the volume through the point `slice_point` of this
```
see comment above

nilsdeppe reviewed (dismissed):
squash

wthrowe committed:
Explicitly instantiate Domain default argument

It seems clang 3.9.1 doesn't implicitly instantiate it, either when
the class is instantiated or when the constructor is called without
specifying the argument.  (Explicitly passing the default value
worked.)
